File: Crypto/Cipher/Threefish/UBI.hs
-- | Tweak manipulation for Unique Block Iteration mode.
module Crypto.Cipher.Threefish.UBI where
import Data.Word
import Data.Bits
import Crypto.Cipher.Threefish.Common

data BlockType
  = Key
  | Config
  | Personalization
  | PublicKey
  | KeyIdentifier
  | Nonce
  | Message
  | Output

type2w64 :: BlockType -> Word64
type2w64 Key             = 0
type2w64 Config          = 4
type2w64 Personalization = 8
type2w64 PublicKey       = 12
type2w64 KeyIdentifier   = 16
type2w64 Nonce           = 20
type2w64 Message         = 48
type2w64 Output          = 63

{-# INLINE newTweak #-}
newTweak :: BlockType -> Tweak
newTweak t = setType t $ setFirst True $ Tweak 0 0

{-# INLINE setType #-}
setType :: BlockType -> Tweak -> Tweak
setType t (Tweak lo hi) =
    Tweak lo ((type2w64 t `shiftL` 56) .|. (hi .&. zeroType))
  where
    zeroType = complement (63 `shiftL` 56)

{-# INLINE setFirst #-}
setFirst :: Bool -> Tweak -> Tweak
setFirst set (Tweak lo hi) = Tweak lo ((if set then setBit else clearBit) hi 62)

{-# INLINE setLast #-}
setLast :: Bool -> Tweak -> Tweak
setLast set (Tweak lo hi) = Tweak lo ((if set then setBit else clearBit) hi 63)

{-# INLINE addBytes #-}
addBytes :: Word64 -> Tweak -> Tweak
addBytes bs (Tweak lo hi) = Tweak (lo + bs) hi

{-# INLINE configTweak #-}
configTweak :: Tweak
configTweak = setFirst True $ setLast True $ newTweak Config

File: cbits/ubi.c
#include "threefish.h"

inline void init_tweak(UBIType type, W64* t) {
  t[0] = 0;
  t[1] = 0;
  set_first(1, t);
  set_type(type, t);
}

inline void mk_config_tweak(W64* t) {
  init_tweak(T_CONFIG, t);
  set_last(1, t);
}

inline void set_type(UBIType type, W64* t) {
  t[1] = (((W64)type) << 56) | (t[1] & ~(((W64)63) << 56));
}

inline void set_first(unsigned char first, W64* t) {
  if(first) {
    t[1] |= (((W64)1) << 62);
  } else {
    t[1] &= ~(((W64)1) << 62);
  }
}

inline void set_last(unsigned char last, W64* t) {
  if(last) {
    t[1] |= (((W64)1) << 63);
  } else {
    t[1] &= ~(((W64)1) << 63);
  }
}

inline void add_bytes(W64 bytes, W64* t) {
  t[0] += bytes;
}
